File: python/codeforces/ostap_and_grasshopper/ostap_and_grasshopper.py
from collections import deque

if __name__ == '__main__':
    found = False
    num_cells, jump_length = tuple(map(int, input().strip().split()))
    line = input().strip()
    target = line.find('T')
    visited = set()
    queue = deque()
    queue.appendleft(line)
    while queue:
        line = queue.pop()
        visited.add(line)
        g_idx = line.find('G')
        if g_idx + jump_length == target or g_idx - jump_length == target:
            found = True
            break
        if g_idx + jump_length < len(line) and line[g_idx+jump_length] != '#':
            new_line = list(line)
            new_line[g_idx] = '.'
            new_line[g_idx+jump_length] = 'G'
            new_line = ''.join(new_line)
            if new_line not in visited:
                queue.appendleft(new_line)
        if g_idx - jump_length >= 0 and line[g_idx-jump_length] != '#':
            new_line = list(line)
            new_line[g_idx] = '.'
            new_line[g_idx-jump_length] = 'G'
            new_line = ''.join(new_line)
            if new_line not in visited:
                queue.appendleft(new_line)
    if found:
        print("YES")
    else:
        print("NO")
    

File: cpp/codeforces/ostap_and_grasshopper/ostap_and_grasshopper.cpp
#include <iostream>
#include <cstdio>
#include <string>
#include <unordered_set>
#include <queue>

using namespace std;

int main() {
    bool found = false;
    int num_cells, length_jump;
    scanf("%d %d\n", &num_cells, &length_jump);
    string line;
    getline(cin, line);
    int target = line.find('T');
    queue<string> q;
    unordered_set<string> visited;
    q.push(line);
    while(!q.empty()) {
        string line = q.front();
        q.pop();
        visited.insert(line);
        int g_idx = line.find('G');
        if(g_idx + length_jump == target || g_idx - length_jump == target) {
            found = true;
            break;
        }
        if(g_idx + length_jump < line.size() && line[g_idx+length_jump] != '#') {
            string new_line(line);
            new_line[g_idx] = '.';
            new_line[g_idx + length_jump] = 'G';
            if(visited.find(new_line) == visited.end()) {
                q.push(new_line);
            }
        }
        if(g_idx - length_jump >= 0 && line[g_idx-length_jump] != '#') {
            string new_line(line);
            new_line[g_idx] = '.';
            new_line[g_idx - length_jump] = 'G';
            if(visited.find(new_line) == visited.end()) {
                q.push(new_line);
            }
        }
    }

    if(found) {
        cout << "YES" << endl;
    }
    else {
        cout << "NO" << endl;
    }
}
